Fix Flakey test TestNegotiationNeededRemoveTrack

```
=== RUN   TestNegotiationNeededRemoveTrack
    TestNegotiationNeededRemoveTrack: peerconnection_renegotiation_test.go:866:
                Error Trace:    peerconnection_renegotiation_test.go:866
                                                        peerconnection.go:317
                                                        operations.go:84
                                                        asm_amd64.s:1373
                Error:          Received unexpected error:
                                InvalidStateError: connection closed
                Test:           TestNegotiationNeededRemoveTrack
    TestNegotiationNeededRemoveTrack: peerconnection_renegotiation_test.go:869:
                Error Trace:    peerconnection_renegotiation_test.go:869
                                                        peerconnection.go:317
                                                        operations.go:84
                                                        asm_amd64.s:1373
                Error:          Received unexpected error:
                                InvalidStateError: connection closed
                Test:           TestNegotiationNeededRemoveTrack
    TestNegotiationNeededRemoveTrack: peerconnection_renegotiation_test.go:872:
                Error Trace:    peerconnection_renegotiation_test.go:872
                                                        peerconnection.go:317
                                                        operations.go:84
                                                        asm_amd64.s:1373
                Error:          Received unexpected error:
                                InvalidStateError: connection closed
                Test:           TestNegotiationNeededRemoveTrack
    TestNegotiationNeededRemoveTrack: peerconnection_renegotiation_test.go:875:
                Error Trace:    peerconnection_renegotiation_test.go:875
                                                        peerconnection.go:317
                                                        operations.go:84
                                                        asm_amd64.s:1373
                Error:          Received unexpected error:
                                InvalidStateError: connection closed
                Test:           TestNegotiationNeededRemoveTrack
    TestNegotiationNeededRemoveTrack: peerconnection_renegotiation_test.go:878:
                Error Trace:    peerconnection_renegotiation_test.go:878
                                                        peerconnection.go:317
                                                        operations.go:84
                                                        asm_amd64.s:1373
                Error:          Received unexpected error:
                                InvalidStateError: connection closed
                Test:           TestNegotiationNeededRemoveTrack
    TestNegotiationNeededRemoveTrack: peerconnection_renegotiation_test.go:881:
                Error Trace:    peerconnection_renegotiation_test.go:881
                                                        peerconnection.go:317
                                                        operations.go:84
                                                        asm_amd64.s:1373
                Error:          Received unexpected error:
                                InvalidStateError: connection closed
                Test:           TestNegotiationNeededRemoveTrack
panic: sync: negative WaitGroup counter
```

--- a/peerconnection.go
+++ b/peerconnection.go
@@ -48,7 +48,7 @@ type PeerConnection struct {
 	idpLoginURL *string
 
 	isClosed               *atomicBool
-	negotiationNeeded      bool
+	isNegotiationNeeded    *atomicBool
 	negotiationNeededState negotiationNeededState
 
 	lastOffer  string
@@ -105,7 +105,7 @@ func (api *API) NewPeerConnection(configuration Configuration) (*PeerConnection,
 		},
 		ops:                    newOperations(),
 		isClosed:               &atomicBool{},
-		negotiationNeeded:      false,
+		isNegotiationNeeded:    &atomicBool{},
 		negotiationNeededState: negotiationNeededStateEmpty,
 		lastOffer:              "",
 		lastAnswer:             "",
@@ -262,8 +262,7 @@ func (pc *PeerConnection) onNegotiationNeeded() {
 	if pc.negotiationNeededState == negotiationNeededStateRun {
 		pc.negotiationNeededState = negotiationNeededStateQueue
 		return
-	}
-	if pc.negotiationNeededState == negotiationNeededStateQueue {
+	} else if pc.negotiationNeededState == negotiationNeededStateQueue {
 		return
 	}
 
@@ -301,19 +300,17 @@ func (pc *PeerConnection) negotiationNeededOp() {
 
 	// Step 2.4
 	if !pc.checkNegotiationNeeded() {
-		pc.negotiationNeeded = false
+		pc.isNegotiationNeeded.set(false)
 		return
 	}
 
 	// Step 2.5
-	if pc.negotiationNeeded {
+	if pc.isNegotiationNeeded.get() {
 		return
 	}
 
 	// Step 2.6
-	pc.mu.Lock()
-	pc.negotiationNeeded = true
-	pc.mu.Unlock()
+	pc.isNegotiationNeeded.set(true)
 
 	// Step 2.7
 	if pc.onNegotiationNeededHandler != nil {
@@ -894,9 +891,7 @@ func (pc *PeerConnection) setDescription(sd *SessionDescription, op stateChangeO
 	if err == nil {
 		pc.signalingState.Set(nextState)
 		if pc.signalingState.Get() == SignalingStateStable {
-			pc.mu.Lock()
-			pc.negotiationNeeded = false
-			pc.mu.Unlock()
+			pc.isNegotiationNeeded.set(false)
 			pc.onNegotiationNeeded()
 		}
 		pc.onSignalingStateChange(nextState)
@@ -1749,15 +1744,15 @@ func (pc *PeerConnection) WriteRTCP(pkts []rtcp.Packet) error {
 
 // Close ends the PeerConnection
 func (pc *PeerConnection) Close() error {
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #2)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #1)
 	if pc.isClosed.get() {
 		return nil
 	}
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #3)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #2)
 	pc.isClosed.set(true)
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #4)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #3)
 	pc.signalingState.Set(SignalingStateClosed)
 
 	// Try closing everything and collect the errors
@@ -1768,14 +1763,14 @@ func (pc *PeerConnection) Close() error {
 	//    continue the chain the Mux has to be closed.
 	closeErrs := make([]error, 4)
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #5)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #4)
 	for _, t := range pc.GetTransceivers() {
 		if !t.stopped {
 			closeErrs = append(closeErrs, t.Stop())
 		}
 	}
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #6)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #5)
 	if pc.sctpTransport != nil {
 		pc.sctpTransport.lock.Lock()
 		for _, d := range pc.sctpTransport.dataChannels {
@@ -1784,20 +1779,20 @@ func (pc *PeerConnection) Close() error {
 		pc.sctpTransport.lock.Unlock()
 	}
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #7)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #6)
 	if pc.sctpTransport != nil {
 		closeErrs = append(closeErrs, pc.sctpTransport.Stop())
 	}
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #8)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #7)
 	closeErrs = append(closeErrs, pc.dtlsTransport.Stop())
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #9,#10,#11)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #8, #9, #10)
 	if pc.iceTransport != nil {
 		closeErrs = append(closeErrs, pc.iceTransport.Stop())
 	}
 
-	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #12)
+	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #11)
 	pc.updateConnectionState(pc.ICEConnectionState(), pc.dtlsTransport.State())
 
 	return util.FlattenErrs(closeErrs)

--- a/peerconnection_renegotiation_test.go
+++ b/peerconnection_renegotiation_test.go
@@ -806,6 +806,7 @@ func TestNegotiationTrackAndChannel(t *testing.T) {
 	})
 
 	pcOffer.OnNegotiationNeeded(func() {
+		wg.Add(1)
 		<-isMulti
 		offer, err := pcOffer.CreateOffer(nil)
 		assert.NoError(t, err)
@@ -830,6 +831,7 @@ func TestNegotiationTrackAndChannel(t *testing.T) {
 		if err = pcOffer.SetRemoteDescription(*pcAnswer.LocalDescription()); err != nil {
 			t.Error(err.Error())
 		}
+		wg.Done()
 		wg.Done()
 	})
 
@@ -862,6 +864,7 @@ func TestNegotiationNeededRemoveTrack(t *testing.T) {
 	assert.NoError(t, err)
 
 	pcOffer.OnNegotiationNeeded(func() {
+		wg.Add(1)
 		offer, createOfferErr := pcOffer.CreateOffer(nil)
 		assert.NoError(t, createOfferErr)
 
@@ -879,6 +882,7 @@ func TestNegotiationNeededRemoveTrack(t *testing.T) {
 
 		<-answerGatheringComplete
 		assert.NoError(t, pcOffer.SetRemoteDescription(*pcAnswer.LocalDescription()))
+		wg.Done()
 		wg.Done()
 	})
 
@@ -911,16 +915,15 @@ func TestNegotiationNeededStressOneSided(t *testing.T) {
 	api.mediaEngine.RegisterDefaultCodecs()
 	pcA, pcB, err := api.newPair(Configuration{})
 	assert.NoError(t, err)
-	defer pcA.Close()
-	defer pcB.Close()
 
+	var wg sync.WaitGroup
 	pcA.OnNegotiationNeeded(func() {
+		wg.Add(1)
 		assert.NoError(t, signalPair(pcA, pcB))
+		wg.Done()
 	})
 
 	for i := 0; i < 500; i++ {
-		time.Sleep(10 * time.Millisecond)
-
 		track, err := pcA.NewTrack(DefaultPayloadTypeVP8, rand.Uint32(), "video", "pion")
 		assert.NoError(t, err)
 
@@ -929,9 +932,15 @@ func TestNegotiationNeededStressOneSided(t *testing.T) {
 
 		err = track.WriteSample(media.Sample{Data: []byte{0x00}, Samples: 1})
 		assert.NoError(t, err)
+
+		time.Sleep(10 * time.Millisecond)
 	}
 
-	// Wait for async to complete otherwise dangling renegotation
-	// will occur on closed connection and throw InvalidStateError: connection closed
-	time.Sleep(time.Duration(2) * time.Second)
+	// Wait for goroutines triggered by `WriteSample` to execute.
+	time.Sleep(100 * time.Millisecond)
+
+	wg.Wait()
+
+	assert.NoError(t, pcA.Close())
+	assert.NoError(t, pcB.Close())
 }

